library/src-bootstrapped/scala/runtime/quoted/Matcher.scala

```diff
@@ -0,0 +1,276 @@
+package scala.runtime.quoted
+
+import scala.annotation.internal.sharable
+
+import scala.quoted._
+import scala.tasty._
+
+object Matcher {
+
+  private final val debug = false
+
+  /**
+   *
+   * @param scrutineeExpr
+   * @param patternExpr
+   * @param reflection
+   * @return None if it did not match, Some(tup) if it matched where tup contains Expr[_], Type[_] or Binding[_]
+   */
+  def unapply[Tup <: Tuple](scrutineeExpr: Expr[_])(implicit patternExpr: Expr[_], reflection: Reflection): Option[Tup] = {
+    import reflection._
+
+    def treeMatches(scrutinee: Tree, pattern: Tree)(implicit env: Set[(Symbol, Symbol)]): Option[Tuple] = {
+
+      /** Check that both are `val` or both are `lazy val` or both are `var` **/
+      def checkValFlags(): Boolean = {
+        import Flags._
+        val sFlags = scrutinee.symbol.flags
+        val pFlags = pattern.symbol.flags
+        sFlags.is(Lazy) == pFlags.is(Lazy) && sFlags.is(Mutable) == pFlags.is(Mutable)
+      }
+
+      def isMatcherHole(sym: Symbol): Boolean =
+        sym.owner.fullName == "scala.runtime.quoted.Matcher$" // TODO check symbol equality instead of its name
+
+      def treesMatch(scrutinees: List[Tree], patterns: List[Tree]): Option[Tuple] =
+        if (scrutinees.size != patterns.size) None
+        else foldMatchings(scrutinees.zip(patterns).map(treeMatches): _*)
+
+//      def bindingAndTptMatch(sym: Symbol, tpt1: TypeTree, tpt2: TypeTree): (Option[Tuple], Option[Tuple]) = tpt2 match {
+//        case bindHole @ TypeTree.Applied(TypeIdent("bindHole"), IsTypeTree(tpt2) :: Nil)
+//            if isMatcherHole(bindHole.symbol) && tpt1.tpe <:< tpt2.tpe => // Is the subtype check required?
+//          val binding = new Binding(sym.name, sym)
+//          (Some(Tuple1(binding)), treeMatches(tpt1, tpt2))
+//        case returnTpt2 =>
+//          (Some(()), treeMatches(tpt1, tpt2))
+//      }
+
+      (scrutinee, pattern) match {
+        // Normalize blocks without statements
+        case (Block(Nil, expr), _) => treeMatches(expr, pattern)
```

Comment:
Instead of doing the simplifications (twice) in the match, maybe apply a `normalize` function beforehand? I.e.
```
   def normalize(tree: Tree) = tree match {
     case Block(Nil, pat) => pat
     case Inlined...
```
Not sure whether it's worth it if it's just those two, though.

Reply:
I added it. We may have some other normalization on blocks such as flattening.